yksgi/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+    }
+}
+
+/// A frame holding live variables.
+struct Frame {
+    vars: HashMap<LLVMValue, SGValue>,
+}
+
+impl Frame {
+    fn new() -> Frame {
+        Frame {
+            vars: HashMap::new(),
+        }
+    }
+
+    /// Get the value of the variable `key` in this frame.
+    fn get(&self, key: &LLVMValue) -> Option<&SGValue> {
+        self.vars.get(key)
+    }
+
+    /// Insert new variable into this frame.
+    fn add(&mut self, key: LLVMValue, val: SGValue) {
+        if let Some(entry) = self.vars.get_mut(&key) {
+            *entry = val;
+        } else {
+            self.vars.insert(key, val);
+        }
+    }
+}
+
+/// The stopgap interpreter. Used during guard failures to get back to the control point by
+/// interpreting LLVM IR.
+pub struct SGInterp {
+    /// LLVM IR module we are interpreting.
+    module: LLVMModule,
+    /// Current frames.
+    frames: Vec<Frame>,
+    /// Current instruction being interpreted.
+    pc: LLVMValue,
+    /// Last basic block that was interpreted.
+    lastbb: Option<LLVMBasicBlock>,
+}
+
+impl SGInterp {
+    /// Create a new stopgap interpreter and initialise it to start interpretation at the location
+    /// given by a basic block index, instruction index, and function name.
+    /// FIXME: Support initialisation of multiple frames.
+    pub unsafe fn new(bbidx: usize, instridx: usize, fname: &CStr) -> SGInterp {
+        // Get AOT module IR and parse it.
+        let module = LLVMModule::from_bc();
+        // Create and initialise stop gap interpreter.
+        let func = module.function(fname.as_ptr());
+        let bb = func.bb(bbidx);
+        let instr = bb.instruction(instridx);
+        SGInterp {
+            module,
+            frames: vec![Frame::new()],
+            pc: instr,
+            lastbb: None,
+        }
+    }
+
+    /// Add a live variable and its value to the current frame.
+    pub unsafe fn init_live(&mut self, bbidx: usize, instridx: usize, fname: &CStr, val: u64) {
+        let func = self.module.function(fname.as_ptr());
+        let bb = func.bb(bbidx);
+        let instr = bb.instruction(instridx);
+        let orgaot = get_aot_original(&instr);
+        let ty = instr.get_type();
+        let value = SGValue::new(val, ty);
+        self.frames.last_mut().unwrap().add(instr, value);
+        if let Some(v) = orgaot {
+            self.frames.last_mut().unwrap().add(v, value);
+        }
+    }
+
+    /// Lookup the value of variable `var` in the current frame.
+    unsafe fn lookup(&self, var: &LLVMValue) -> SGValue {
+        if var.is_instruction() {
+            *self.frames.last().unwrap().get(var).unwrap()
+        } else if var.is_constant() {
+            llvmbridge::llvm_const_to_sgvalue(var.0)
+        } else {
+            // GlobalVariable, Function, etc.
+            todo!()
+        }
+    }
+
+    /// Lookup the value of variable `var` in the current frame.
+    unsafe fn setvar(&mut self, var: LLVMValue, val: SGValue) {
```

Comment:
Can we use consistent names for lookup/setvar? `var_set` and `var_lookup` are often good bets in my experience!

Reply:
Sounds good.

Reply:
Fixed in 710be13